### ISSUE TITLE
[FEA] Groupby normalize floating point

### ISSUE THREAD

**Issue by abellina** [rapidsai/cudf#2500]
Spark expects normalized floating point values when grouping. For example:

`-0.0` should be treated as `0.0`, and 
`NaN` should be equal to `NaN`. 

This feature request is to add a way to transform an entry of a column before we group on it, s.t. we can emulate the same behavior. We can work around this for now, by using `find_and_replace_all`, but it would be better if the groupby/aggregate could handle it.

**Comment by abellina** [comment]
@jrhemstad

**Comment by jrhemstad** [comment]
It's actually even more involved than what you're requesting. The best way you could emulate this today is to convert a `NaN` to a `NULL` with [`nans_to_nulls`](https://github.com/rapidsai/cudf/pull/2358).

By definition of the IEEE 754 standard, `x == NaN` is always false for all values of `x` (including `NaN`). 

Therefore, in order to be able to group `NaN`s together, when comparing `a == b`, we'd have to _specifically_ check for `if(is_nan(a) and is_nan(b) return true;`.

**Comment by abellina** [comment]
Yeah agree, I see there is some work on this already, so I am adding in this issue some examples as to what Spark does with null and NaN.

Create a test dataframe:
```
var df = Seq[(java.lang.Integer, java.lang.Float, java.lang.Double)](
      (0, 0.0f, 0.0),
      (-0, -0.0f, -0.0),
      (1, 1.0f, Double.NaN),
      (1, 2.0f, Double.NaN),
      (2, Float.NaN, 2.0),
      (2, Float.NaN, 2.0),
      (3, 3.0f, 3.0),
      (3, 3.0f, 3.0),
      (4, 4.0f, null),
      (4, 4.0f, null),
      (5, null, 5.0),
      (5, null, 5.0),
      (null, 6.0f, 6.0),
      (null, 6.0f, 6.0),
    ).toDF("ints", "floats", "doubles")
df.createOrReplaceTempView("test")
```
Group by ints. Please note that:
- `null == null` in group by
- `-0 == 0` in Java already, so they are grouped
- skips nulls in this case for the count(col), but not for count(*)
- doesn't skip NaN.
```
scala> spark.sql("select ints, count(*), count(floats), count(doubles) from test group by ints").show
+----+--------+-------------+--------------+
|ints|count(1)|count(floats)|count(doubles)|
+----+--------+-------------+--------------+
|null|       2|            2|             2|
|   1|       2|            2|             2|
|   3|       2|            2|             2|
|   5|       2|            0|             2|
|   4|       2|            2|             0|
|   2|       2|            2|             2|
|   0|       2|            2|             2|
+----+--------+-------------+--------------+
```
Floats
Group by:
- `null == null`
- `-0.0f == 0.0f`, which is normalized in Spark not in Java.
- `NaN == NaN`

Aggregation:
- skips nulls in this case for the count(col), but not for count(*)
- doesn't skip NaN.
```
scala> spark.sql("select floats, count(*), count(ints), count(doubles) from test group by floats").show
+------+--------+-----------+--------------+
|floats|count(1)|count(ints)|count(doubles)|
+------+--------+-----------+--------------+
|  null|       2|          2|             2|
|   2.0|       1|          1|             1|
|   3.0|       2|          2|             2|
|   1.0|       1|          1|             1|
|   6.0|       2|          0|             2|
|   4.0|       2|          2|             0|
|   NaN|       2|          2|             2|
|   0.0|       2|          2|             2|
+------+--------+-----------+--------------+
```
Doubles (*this is exactly the same behavior as Float*)
Group by:
- `null == null`
- `-0.0f == 0.0f`, which is normalized in Spark not in Java.
- `NaN == NaN`

Aggregation:
- skips nulls in this case for the count(col), but not for count(*)
- doesn't skip NaN.
```
scala> spark.sql("select doubles, count(*), count(ints), count(floats) from test group by doubles").show
+-------+--------+-----------+-------------+
|doubles|count(1)|count(ints)|count(floats)|
+-------+--------+-----------+-------------+
|    0.0|       2|          2|            2|
|    NaN|       2|          2|            2|
|   null|       2|          2|            2|
|    3.0|       2|          2|            2|
|    2.0|       2|          2|            2|
|    6.0|       2|          0|            2|
|    5.0|       2|          2|            0|
+-------+--------+-----------+-------------+
```

**Comment by jrhemstad** [comment]
I think the better solution to this is to specialize the libcudf hash functions such that all the types of `NaN` values hash to the same hash value. Would that suffice instead of needing to run a kernel to normalize all nans?

**Comment by jrhemstad** [comment]
See https://github.com/rapidsai/cudf/issues/3233

**Comment by jlowe** [comment]
One issue with implicitly converting in the hash/compare functions is what key value will be representative of a group.  For example, if grouping by a key column that contains both -NaN and NaN, would the resulting group have a key value of -NaN or NaN?  In Spark this would always be NaN since it explicitly normalizes -0 to 0 and -NaN to NaN before the groupby.  If we fix this up implicitly in the hashing operation then I suspect the resulting group key value could be either -NaN or NaN in a non-deterministic manner.

I don't know if this difference in behavior would matter in practice. However if we want to err on the side of caution, I think the separate conversion function would be useful.  Spark explicitly calls out the normalization in its query plans, so it is straightforward to map if there's a normalization function in libcudf.

**Comment by harrism** [comment]
Closed by #3291 .
